Partial replication changes

arj03 opened the pull request:
WIP

staltz commented on test/clock.js:
```diff
@@ -0,0 +1,74 @@
+  t.equal(clockAliceAfter[alice.id], 1, 'clock ok')
+  t.equal(clockAliceAfter[bob.id], 1, 'clock ok')
+
+  const clockBobAfter = await pify(alice.ebt.clock)()
```
```suggestion
  const clockBobAfter = await pify(bob.ebt.clock)()
```

staltz commented on test/clock.js:
```diff
@@ -0,0 +1,74 @@
+  const clockAliceAfter2 = await pify(alice.ebt.clock)()
+  t.equal(clockAliceAfter2[alice.id], 2, 'clock ok')
+
+  const clockBobAfter2 = await pify(alice.ebt.clock)()
```
```suggestion
  const clockBobAfter = await pify(bob.ebt.clock)()
```

staltz commented on index.js:
```diff
@@ -23,7 +23,8 @@ exports.manifest = {
   block: 'sync',
-  peerStatus: 'sync'
+  peerStatus: 'sync',
+  getClock: 'async'
```
`getClock` in the manifest but the method exported is `clock`. Also we might need this added to `permissions`/`anonymous`?

arj03 commented:
Thanks @staltz. I just pushed up what I had to get a better overview. I'm writing more tests to see if this works properly for bendy butt :)

staltz commented:
Excited to see this. I know you're not asking for reviews yet, I'm just highlighting some typos that seem obvious

arj03 commented:
Thanks, appreciated :)

arj03 commented:
There is something fishy going on in db2/compat/ebt.js + this depends on https://github.com/ssbc/epidemic-broadcast-trees/pull/50

arj03 commented:
All green :)

staltz commented on index.js:
```diff
@@ -275,7 +281,8 @@ exports.init = function (sbot, config) {
+    replicateFormat,
     peerStatus,
     clock,
     registerFormat
```
duplicate down here

arj03 replied:
That's registerFormat :)

staltz replied:
Ah, my bad my bad

staltz commented:
Probably this was discussed in your recent video call, but what's the rationale behind introducing `replicateFormat`?

arj03 commented:
The rationale behind `replicateFormat` is that we were thinking about how to handle "old" EBT clients that only support classic replication. We didn't want to change the version number as that is more tied to the wire format than the options and would break old clients. So we opted to use replicate for classic feed formats and a new `replicateFormat` for other feed formats, that way replication with existing client still works fine.

cryptix commented on index.js:
```diff
@@ -44,48 +46,105 @@ function cleanClock (clock) {
+  function getEBT(formatName) {
+    const ebt = ebts[formatName]
+    if (!ebt)
+      throw new Error('Unknown format' + formatName)
```
```suggestion
      throw new Error('Unknown format: ' + formatName)
```

arj03 replied:
Thanks :smiley:

arj03 commented:
@cryptix I changed the clock to take an object instead to better match the other APIs and make it more future-proof

cryptix commented on test/formats.js:
```diff
@@ -0,0 +1,480 @@
+  keys: u.keysFor('bob')
+})
+
+const bendyButtMethods = {
```
I copied this object to my test now but that feels wrong..

Could we maybe expose this via `require('ssb-bendy-butt/ebt')` or something like this?

cryptix replied:
or `ssb-ebt/bendy-butt` what ever you feel is best for maintenance. :)

arj03 replied:
Right, that seems like a good idea. Similar with indexed thing. André is working on changing ssb-meta-feeds so we can get the information on what feeds are indexes and use that instead of adding the author to index among other things :)

cryptix commented on test/formats.js:
```diff
@@ -0,0 +1,480 @@
+let bobMFId
+
+tape('multiple formats', async (t) => {
+  alice.ebt.registerFormat('bendybutt', bendyButtMethods)
```
Can we use the constants from ssb-uri here?

so `bendybutt-v1` instead of this and `ed25519` instead of `classic`?

arj03 replied:
Good idea, thanks for the input :)

staltz replied:
@arj03 If it helps, I can start renaming `classic` to `ed25519` in ssb-meta-feeds and downstream

staltz replied:
Aaactually, ssb-bfe-spec relies on `classic` https://github.com/ssb-ngi-pointer/ssb-bfe-spec/blob/a4e38c30b67d7110115c7fc8440b71448f87d8d2/bfe.json#L6 so we could keep it `classic` while rename `bendybutt` to `bendybutt-v1`.

arj03 replied:
Personally I prefer the `classic` name :)

staltz commented:
ssb-meta-feeds could be updated to 0.20.0

staltz commented:
@arj03 Note I added two simple commits to this branch, before you bump into local branch collisions

staltz commented on formats/indexed.js:
```diff
@@ -0,0 +1,45 @@
+          if (err) return cb(err)
+          const { sequence } = msg.value.content.indexed
+          const authorInfo = sbot.metafeeds.findByIdSync(msg.value.author)
+          if (!authorInfo) return cb(new Error("Unknown author", msg.value.author))
```
`"Unknown author: " + msg.value.author`

Note that if you use comma in an Error constructor, the 2nd arg is actually "filename" 🤯

staltz commented on formats/indexed.js:
```diff
@@ -0,0 +1,45 @@
+          const { sequence } = msg.value.content.indexed
+          const authorInfo = sbot.metafeeds.findByIdSync(msg.value.author)
+          if (!authorInfo) return cb(new Error("Unknown author", msg.value.author))
+          const { author } = JSON.parse(authorInfo.metadata.query)
```
We could use QL0.parse here, from ssb-subset-ql, which also does validation. might be useful to handle those possible errors

staltz commented on formats/sliced.js:
```diff
@@ -0,0 +1,12 @@
+const classic = require('./classic')
+
+module.exports = Object.assign(
+  {}, classic, {
```
Object spread

staltz reviewed (changes requested):
Did a quick review on mobile. Will take another look again today

arj03 commented on README.md:
````diff
@@ -96,16 +96,94 @@ The output looks like this:
+```js
+{
+  // used in request, block, cleanClock, sbot.post, vectorClock
+  sbotIsFeed(sbot, feedId) {
````
That's for this change in index.js:

```
// EBT expects a function of only feedId so we bind sbot here
format.isFeed = (feedId) => format.sbotIsFeed(sbot, feedId)
```

Very open if you have a better way to solve that. The also means that the methods now bind sbot so can't be shared, which is why formats/indexed is a function.

staltz commented on index.js:
```diff
@@ -44,48 +45,85 @@ function cleanClock (clock) {
+
+    const format = formats[formatName]
+    // EBT expects a function of only feedId so we bind sbot here
+    format.isFeed = (feedId) => format.sbotIsFeed(sbot, feedId)
```
To solve the naming issue, you could use `fn.bind(thisArg, arg1, arg2, ...)` like this:

```js
function foo(a,b,c) { console.log(a,b,c) }

//                       this, a
const coolFoo = foo.bind(null, 10)

coolFoo('x','y') // 10 x y
```

So
```suggestion
    format.isFeed = format.isFeed.bind(format, sbot)
```

arj03 replied:
Right, just what I was looking for. Thanks!

staltz commented on README.md:
```diff
@@ -96,16 +96,94 @@ The output looks like this:
 
+By registering a format you create a new EBT instance used for
+replicating feeds using that format. This means its own clock. Message
+will be replicated using the `replicateFormat` api. `formatName` must
```
```suggestion
will be replicated using the `replicateFormat` API. `formatName` must
```

staltz commented on README.md:
```diff
@@ -96,16 +96,94 @@ The output looks like this:
+replicating feeds using that format. This means its own clock. Message
+will be replicated using the `replicateFormat` api. `formatName` must
+be a string and methods must implement the following functions. The
+example shows the 'classic' implementation.
```
```suggestion
example below shows the implementation for 'classic' ed25519 SSB feeds.
```

staltz commented on index.js:
```diff
@@ -44,48 +45,85 @@ function cleanClock (clock) {
-    isFeed: isFeed
-  })
+  const formats = {
+    'classic': require('./formats/classic')
```
How about this object is empty and we delete line 95, and then we just call `registerFormat('classic', classicFormat)`. This would make sure that if we need to update in the future the internals of `registerFormat`, then we also correctly treat the registration of 'classic' automatically.

staltz commented on index.js:
```diff
@@ -44,48 +45,85 @@ function cleanClock (clock) {
+    // EBT expects a function of only feedId so we bind sbot here
+    format.isFeed = (feedId) => format.sbotIsFeed(sbot, feedId)
+
+    const ebt = EBT(Object.assign({
```
Object spread would be good here too

staltz commented on index.js:
```diff
@@ -44,48 +45,85 @@ function cleanClock (clock) {
+  }
+
+  const ebts = {}
+  function addEBT(formatName) {
```
Would be rather easy to do `addEBT(formatName, methods)` or `addEBT(formatName, format)` here and then we don't need to use the global `formats` here. It's good to reduce usage of globals if it's easy, makes this function more refactorable/movable in the future.

arj03 replied:
Yeah this is a good idea, latest changes hopefully makes these things clearer

staltz commented on index.js:
```diff
@@ -44,48 +45,85 @@ function cleanClock (clock) {
+
+        validClock = {}
+        for (let k in clock)
+          if (format.isFeed(k))
```
Is it possible to just use `ebt.isFeed(k)` here? (Not sure if I understood the `EBT()` constructor correctly or not)

staltz commented on index.js:
```diff
@@ -104,29 +143,56 @@ exports.init = function (sbot, config) {
+
+      const format = formats[formatName]
+
+      if (!(format && format.isFeed(destFeedId))) return
```
Would be nice if this one were like lines 190,191,192 in block(). I've always found the obligatory mental De Morgan's law application a bit hard to read with these kinds of `if`s

staltz commented on index.js:
```diff
@@ -139,11 +205,14 @@ exports.init = function (sbot, config) {
       throw new Error('expected ebt.replicate({version: 3})')
     }
 
+    let formatName = opts.format || 'classic'
```
```suggestion
    const formatName = opts.format || 'classic'
```

staltz commented on index.js:
```diff
@@ -104,29 +143,56 @@ exports.init = function (sbot, config) {
+      formatName = 'classic'
+      for (let format in ebts)
+        if (formats[format].isFeed(origFeedId))
+          formatName = format
```
Okay I've seen the 4 lines above copy-pasted in many places so it seems we could make a helper function, called `detectFormatName(feedId)`. Also I noticed that the `formatName` var wasn't declared (not `var formatName`, not `let formatName`) which makes it a global and that's a bit nasty

staltz commented:
@arj03 Also remember to add the `formats` folder to package.json `files` array

staltz commented on formats/indexed.js:
```diff
@@ -0,0 +1,45 @@
+const pify = require('promisify-4loc')
+const { QL0 } = require('ssb-subset-ql')
+
+module.exports = function () {
```
All the other formats are objects, but this one is a function that returns an object. Could this be changed to object? I tripped over it when using this latest branch in replication-scheduler

arj03 commented:
There is one more thing I'd like to do. Currently if you register indexed then it takes over all classic formats. And I'd like it to only handle the ones you specially register.

staltz commented on index.js:
```diff
@@ -161,35 +165,28 @@ exports.init = function (sbot, config) {
   })
 
+  function findEBTForFeed(feedId) {
+    let ebt = Object.values(ebts).reverse().find(ebt => ebt.isFeed(feedId))
```
The reverse() looks funny, I wonder if the next developer will find it confusing. Maybe we could use the previous `for`-loop we had before? That one was pretty clear to me.

staltz replied:
PS: good idea to drop the `formats` global and use just `ebts`

staltz commented:
Question: why do we need to pass a name when registering a format? That name could be anything (as far as I see), so why not let the internals assign any random name? Or even better, use an array of formats instead of a dict?

arj03 commented:
I think that was mostly to make it easier to debug by being able to print out the name. Adding a name attribute to the format object would serve a similar purpose and make things simpler by just having an array.

staltz commented:
Yes! Name field on the format object

arj03 commented:
Some comments on the last changes:

I changed the structure so that ebts are an array now with names. The API seems better, less possibility to do something wrong :)

I changed the sliced example because I found out that in order for the existing to work, then both ends needs to have a feed registered for sliced replication over a new EBT. That's a bit of a tall order, so instead what I did was to change one end (carol) so that she is responsible of what needs to be sliced replicated and what needs to be replicated in full. This way alice doesn't need to know. This is also why I didn't put that sliced replication in formats.

Last I changed `request` and `block` so that you can provide the exact ebt where this action needs to happen. Lets say you have a pub and you want do enable indexed replication, but you don't yourself want to do indexed replication. You instead want to replicate the indexes as normal feeds. Before that was no way to do this, a feed would either be in indexed or classic. With the ability to specify the format this is now possible. I changed the default to be first ebt that matches isFeed. This means that for the indexed example you have to provide the formatName (see test/formats).

staltz commented on formats/bendy-butt.js:
```diff
@@ -0,0 +1,51 @@
+module.exports = {
+  name: 'bendybutt-v1',
+  // used in request, block, cleanClock, sbot.post, vectorClock
+  sbotIsFeed(sbot, feedId) {
```
@arj03 This couldn't be renamed to `isFeed`? Did the `bind` approach not work?

staltz commented on index.js:
```diff
@@ -44,48 +45,96 @@ function cleanClock (clock) {
-    isFeed: isFeed
-  })
+  const ebts = []
+  registerFormat(require('./formats/classic'))
```
Could we put all the requires at the top of the file?

staltz commented on index.js:
```diff
@@ -44,48 +45,96 @@ function cleanClock (clock) {
+    ebt.name = format.name
+
+    const existingId = ebts.findIndex(e => e.name === format.name)
+    if (existingId != -1)
```
```suggestion
    if (existingId !== -1)
```

staltz reviewed (changes requested):
Just 3 comments, otherwise all the rest is good to go :)

Oh, and the very last thing to do is run `npm run lint`

staltz commented on index.js:
```diff
@@ -104,29 +154,56 @@ exports.init = function (sbot, config) {
+    if (formatName)
+      ebt = ebts.find(ebt => ebt.name === formatName)
+    else
+      ebt = ebts.find(ebt => ebt.isFeed(feedId))
```
By the way, when using this latest branch in ssb-replication-scheduler, I keep getting `classic` when the input is an index feed, but if I add `.filter(ebt => ebt.name !== 'classic')` then that fixes it.

staltz commented on index.js:
```diff
@@ -104,29 +151,52 @@ exports.init = function (sbot, config) {
-  function request (destFeedId, requesting) {
+  function findEBTForFeed (feedId, formatName) {
+    let ebt
+    if (formatName) { ebt = ebts.find(ebt => ebt.name === formatName) } else { ebt = ebts.find(ebt => ebt.isFeed(feedId)) }
```
I put a comment here but it got swallowed after the lint (PS: what is the linter doing omg)

> By the way, when using this latest branch in ssb-replication-scheduler, I keep getting classic when the input is an index feed, but if I add .filter(ebt => ebt.name !== 'classic') then that fixes it.

arj03 replied:
Yeah I'm not a super fan of those 1 liner tuborgs. There are several of them it made. The other one we should discuss :)

staltz commented on index.js:
```diff
@@ -104,80 +156,148 @@ exports.init = function (sbot, config) {
-      if (!isFeed(destFeedId)) return
+      const ebt = findEBTForFeed(destFeedId, formatName)
+
+      if (!ebt.isFeed(destFeedId)) return
```
@arj03 In the case of `formatName` being passed as `'indexed'`, then this line still suffers from that race condition because `isFeed` is going to look up the state from ssb-meta-feeds. I commented it out and thought things would work but still no luck. Might be caused by some other bug somewhere

staltz replied:
> Might be caused by some other bug somewhere

Ah, I think I know what it is. indexed format `isFeed` is called also in cleanClock, and in epidemic-broadcast-tree `request`, so we can't go around the race condition, gotta solve it.

staltz commented:
Hey @mixmix we've been working on this PR together for https://github.com/ssb-ngi-pointer/ssb-replication-scheduler/pull/5 and there's lots of changes going on, I thought it might be good to get a (your) third perspective on it.

Currently, IMO this branch is ready to merge, but we'll test it out in https://github.com/ssb-ngi-pointer/ssb-replication-scheduler/pull/5 for the next couple days (probably merging earliest next week) and see if it's stable and doesn't present new problems.

mixmix commented:
I'm confident in your abilities and care. I'll have a small look but not an
expert in this area

On Fri, 17 Sep 2021, 06:14 André Staltz, ***@***.***> wrote:

> Hey @mixmix <https://github.com/mixmix> we've been working on this PR
> together for ssb-ngi-pointer/ssb-replication-scheduler#5
> <https://github.com/ssb-ngi-pointer/ssb-replication-scheduler/pull/5> and
> there's lots of changes going on, I thought it might be good to get a
> (your) third perspective on it.
>
> Currently, IMO this branch is ready to merge, but we'll test it out in
> ssb-ngi-pointer/ssb-replication-scheduler#5
> <https://github.com/ssb-ngi-pointer/ssb-replication-scheduler/pull/5> for
> the next couple days (probably merging earliest next week) and see if it's
> stable and doesn't present new problems.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ssbc/ssb-ebt/pull/52#issuecomment-921133480>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAUK3HXYQUCDGD25BZ365GLUCIXWRANCNFSM5DP6WUQA>
> .
>

staltz commented:
@arj03 I was working on handling blocks in ssb-replication-scheduler and here are some thoughts about a corner case:

e-b-t's `state` object has `state.peers` which gets updated with the peers connected via `replicateFormat`, which uses `rpc.id` from SHS. So e-b-t is tracking the "follow graph" but index feeds and meta feeds are never "origins" of social graph edges, they are always the destination. 

This means that the current implementation of `block()` here is a bit confusing. Support origFeedId is classic but destFeedId is bendy butt, then the found `ebt` will be the classic, and then `ebt.isFeed(destFeedId)` will be false:

```js
  function block(origFeedId, destFeedId, blocking, formatName) {
    initialized.promise.then(() => {
      const ebt = findEBTForFeed(origFeedId, formatName)
      ebt.prepareForIsFeed(destFeedId, () => {
        if (!ebt.isFeed(origFeedId)) return
        if (!ebt.isFeed(destFeedId)) return

        if (blocking) {
          ebt.block(origFeedId, destFeedId, true)
        } else if (
          ebt.state.blocks[origFeedId] &&
          ebt.state.blocks[origFeedId][destFeedId]
        ) {
          // only update unblock if they were already blocked
          ebt.block(origFeedId, destFeedId, false)
        }
      })
    })
  }
```

mixmix commented:
Ok, read the readme.
My only question is a naive one, but might be useful - is this the right repo for this complexity?

Motivation of that question : I don't know if this is adding a lot more bells and whistles to ebt. If it is should that complexity live here, of be provided in some other way?

arj03 commented:
Right it doesn't work properly now. We should specify in the README that `origFeedId` should be the main id of the blocker. And use isFeed from classic to check `origFeedId` and let `findEBTForFeed` use `destFeedId`. This leaves it the responsibility of the caller to block the same formats that was also requested. What do you think about that?

staltz commented:
> Motivation of that question : I don't know if this is adding a lot more bells and whistles to ebt. If it is should that complexity live here, of be provided in some other way?

@mixmix Good questions, we had some concerns where should the `formats/*.js` files live, it wasn't obvious where to put those. Our best options were this (what we're doing now in this branch) OR new modules such as `ssb-ebt-classic`, `ssb-ebt-bendybutt`, `ssb-ebt-indexed` which would be plugins for ssb-ebt, but it felt like just tucking the files in arbitrary new modules.

On the other hand, ssb-ebt used to be just a light wrapper around epidemic-broadcast-trees, and I think the philosophy is still the same, but now ssb-ebt can handle multiple instances of epidemic-broadcast-trees. I think overall ssb-ebt is small and has a clear responsibility (provide glues between epidemic-broadcast-trees and SSB feeds and database calls), even after this PR.

> And use isFeed from classic to check `origFeedId` and let `findEBTForFeed` use `destFeedId`. This leaves it the responsibility of the caller to block the same formats that was also requested. What do you think about that?

@arj03 Should work, let's try that.

staltz commented:
@arj03 I found another small issue with blocks:

This line https://github.com/ssbc/ssb-ebt/blob/2fe87910ca37a8dc50d6ad7458617885a3f51e48/index.js#L216 is causing a bug, I think the culprit is the next-tick Promise .then() behavior (yay!).

There's a simple test in ssb-replication-scheduler (`test/integration/block3.js`) that hasn't been passing and the cause is that line. If I comment it out, it works. 

I think it's a race condition. Alice's database gets appended with a new block message, which then updates clock in ssb-ebt, and *should* ASAP also call ssb-ebt block() and prevent that message from being sent to Bob (now blocked) but I think the block() is too slow (because of the promise) so Bob ends up getting the message.

Obviously we could fix that by using raw callbacks that don't have forced-next-tick async behavior, but it seems like a gotcha and maybe a better solution is adding an explicit wait mechanism for blocks to be computed before sending new info to remote peers. This way we don't need the block part to run ASAP, it is allowed to be async.

arj03 commented:
Nice find. 

That really seems quite brittle overall, even before. You are talking about changing `sbot.post` to process blocks differently? What I find strange is that they both have this `initialized.promise.then`, it's not the prepareForIsFeed in block? Mostly trying to understand. 

What I mean with brittle is that the post comes from db2 as the message is appended, while the blocking comes from ssb-friends (an index). Adding more of these kind of synchonizations between different modules feels like we are going back into the bermuda again. Hmm.

staltz commented:
> You are talking about changing `sbot.post` to process blocks differently? What I find strange is that they both have this `initialized.promise.then`, it's not the prepareForIsFeed in block? Mostly trying to understand.

Oh yeah, I changed the implementation of block() so that it would do `ebts.forEach(ebt => ebt.block(etc))` (because I was desperate to make it work) and the only thing wrapping that was the promise.then(), when I removed the promise.then(), then the forEach block worked. I assume that the `prepareForIsFeed` could also add a slow async check, but on the other hand I think that `prepareForIsFeed` for the classic is synchronous because it does the cb() call immediately.

> Adding more of these kind of synchonizations between different modules feels like we are going back into the bermuda again. Hmm.

Very true, I didn't think of it like that. Maybe there is another way. Perhaps it's not a showstopper that Bob gets a few messages before Alice's block kicks in. And/Or perhaps it's okay to do a debounce or delay on ssb-ebt `ssb.post` so that local database plugins (like ssb-friends) take priority first.

staltz commented:
Update: another idea is that we put these kind of race condition solutions in ssb-replication-scheduler (which is THE module for gluing all these things together). Maybe we could add an `post`-like `obv` to ssb-ebt and ssb-friends so that ssb-replication-scheduler knows the latest seq of both of them, and then explicitly pauses ssb-ebt before it has caught up with ssb-friends.

arj03 commented:
I like the last solution a lot better 👍

staltz commented:
For that, we will need some way of feeding new db messages to ssb-ebt to replace sbot.post, or we need pause/resume in ssb-ebt

arj03 commented:
For future travelers the problem described above should have been solved in https://github.com/ssbc/ssb-ebt/pull/55

arj03 commented:
Been testing this branch in the 8k demo and it has been working fine. I wonder if we should release this as 8.0 to try and reduce the number of dominoes we have left or we should wait for further testing of the [replication-scheduler PR](https://github.com/ssb-ngi-pointer/ssb-replication-scheduler/pull/5)?

staltz commented:
Hmmm, I feel like we should test a bit more. There's anyway just two branches to babysit: this one and the replication-scheduler, not that many.
